File: src/main/java/bad/robot/radiate/MonitoringTasks.java
package bad.robot.radiate;

import bad.robot.radiate.monitor.Monitor;
import bad.robot.radiate.monitor.MonitoringTask;
import bad.robot.radiate.monitor.MonitoringTasksFactory;
import bad.robot.radiate.monitor.NothingToMonitorException;

import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ScheduledFuture;

import static bad.robot.radiate.RestartRequired.restartRequired;
import static java.util.Collections.emptyList;

public class MonitoringTasks implements Iterable<MonitoringTask> {

    private final Monitor monitor;

    private List<MonitoringTask> tasks = emptyList();
    private Iterable<ScheduledFuture<?>> scheduled = emptyList();

    public MonitoringTasks(MonitoringTasksFactory factory, Monitor monitor) {
        this.monitor = monitor;
        try {
            tasks = factory.create();
        } catch (Exception e) {
            factory.notifyObservers(e);
            factory.notifyObservers(restartRequired());
        } finally {
            if (tasks.isEmpty())
                factory.notifyObservers(new NothingToMonitorException());
        }
    }

    public void start() {
        scheduled = monitor.start(tasks);
    }

    public void stop() {
        monitor.stop(scheduled);
    }

    @Override
    public Iterator<MonitoringTask> iterator() {
        return tasks.iterator();
    }
}

File: src/main/scala/bad/robot/radiate/MonitoringTasks.scala
package bad.robot.radiate

import java.util.concurrent.ScheduledFuture

import bad.robot.radiate.monitor._

class MonitoringTasksS(factory: MonitoringTasksFactory, monitor: Monitor) extends Iterable[MonitoringTask] {

  private var scheduled = List.empty[ScheduledFuture[_]]

  val tasks = factory.create.valueOr(error => {
    factory.notifyObservers(error)
    factory.notifyObservers(RestartRequired())
    List.empty[MonitoringTask]
  })
  if (tasks.isEmpty) factory.notifyObservers(new NothingToMonitorException)

  def start() {
    scheduled = monitor.start(tasks)
  }

  def stop() {
    monitor.cancel(scheduled)
  }

  def iterator: Iterator[MonitoringTask] = tasks.iterator
}